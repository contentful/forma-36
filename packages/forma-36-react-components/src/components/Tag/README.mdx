---
title: 'Tag'
type: 'component'
status: 'stable'
slug: /components/tag
github: 'https://github.com/contentful/forma-36/tree/master/packages/forma-36-react-components/src/components/Tag'
storybook: 'https://f36-storybook.contentful.com/?path=/story/components-tag--basic'
---

Tags are used to label or highlight items in the interface.

## Table of contents

- [How to use Tag](#how-to-use-tag)
- [Component variations](#component-variations)
- [Content recommendations](#content-recommendations)
- [Accessibility](#accessibility)

## How to use Tag

- tags are used to inform users of the status of an element in the interface
- keep in mind the recommended color coding for each of tagType
- there is a `small` size of Tag available. This might be used in denser areas of content

## Component variations

### Tag types

- **Primary** - used for the entity that has been changed, it maps to entityStatusType `changed`

```jsx
<Table>
  <TableHead>
    <TableRow>
      <TableCell>Name</TableCell>
      <TableCell>Content Type</TableCell>
      <TableCell>Author</TableCell>
      <TableCell>Updated</TableCell>
      <TableCell>Status</TableCell>
    </TableRow>
  </TableHead>
  <TableBody>
    <TableRow>
      <TableCell>My entry title</TableCell>
      <TableCell>Blog post</TableCell>
      <TableCell>Jane Roe</TableCell>
      <TableCell>Oct 29, 2020</TableCell>
      <TableCell>
        <Tag tagType="primary">changed</Tag>
      </TableCell>
    </TableRow>
  </TableBody>
</Table>
```

- **Primary filled** - used to highlight a new functionality or to indicate changes in the interface.

```jsx
<FormLabel htmlFor="">
  Instance parameter definitions{' '}
  <Tag size="small" tagType="primary-filled">
    new
  </Tag>
</FormLabel>
```

- **Positive** - used for the entity that has been published, it maps to entityStatusType `published`

```jsx
<Table>
  <TableHead>
    <TableRow>
      <TableCell>Name</TableCell>
      <TableCell>Content Type</TableCell>
      <TableCell>Author</TableCell>
      <TableCell>Updated</TableCell>
      <TableCell>Status</TableCell>
    </TableRow>
  </TableHead>
  <TableBody>
    <TableRow>
      <TableCell>My entry title</TableCell>
      <TableCell>Blog post</TableCell>
      <TableCell>Jane Roe</TableCell>
      <TableCell>Oct 29, 2020</TableCell>
      <TableCell>
        <Tag tagType="positive">published</Tag>
      </TableCell>
    </TableRow>
  </TableBody>
</Table>
```

- **Negative** - used for the entity that has been deleted, it maps to entityStatusType `deleted`.

```jsx
<Table>
  <TableHead>
    <TableRow>
      <TableCell>Name</TableCell>
      <TableCell>Content Type</TableCell>
      <TableCell>Author</TableCell>
      <TableCell>Updated</TableCell>
      <TableCell>Status</TableCell>
    </TableRow>
  </TableHead>
  <TableBody>
    <TableRow>
      <TableCell>My entry title</TableCell>
      <TableCell>Blog post</TableCell>
      <TableCell>Jane Roe</TableCell>
      <TableCell>Oct 29, 2020</TableCell>
      <TableCell>
        <Tag tagType="negative">deleted</Tag>
      </TableCell>
    </TableRow>
  </TableBody>
</Table>
```

- **Warning** - used for the entity that has been deleted, it maps to entityStatusType `deleted`.

```jsx
<Table>
  <TableHead>
    <TableRow>
      <TableCell>Name</TableCell>
      <TableCell>Content Type</TableCell>
      <TableCell>Author</TableCell>
      <TableCell>Updated</TableCell>
      <TableCell>Status</TableCell>
    </TableRow>
  </TableHead>
  <TableBody>
    <TableRow>
      <TableCell>My entry title</TableCell>
      <TableCell>Blog post</TableCell>
      <TableCell>Jane Roe</TableCell>
      <TableCell>Oct 29, 2020</TableCell>
      <TableCell>
        <Tag tagType="warning">draft</Tag>
      </TableCell>
    </TableRow>
  </TableBody>
</Table>
```

- **Secondary** - used for all the secondary elements, for example in the list of versions:

```jsx
<>
  <div>
    <Flex justifyContent="space-between" alignItems="center">
      <Flex
        htmlTag="label"
        paddingTop="spacingXs"
        paddingBottom="spacingXs"
        alignItems="center"
      >
        <RadioButtonField
          type="radio"
          name="selected"
          role="option"
          labelText="Yesterday, 1:36PM"
          checked
        />
      </Flex>
      <div>
        <Tag tagType="secondary">current</Tag>
      </div>
    </Flex>
  </div>
  <div>
    <Flex justifyContent="space-between" alignItems="center">
      <Flex
        htmlTag="label"
        paddingTop="spacingXs"
        paddingBottom="spacingXs"
        alignItems="center"
      >
        <RadioButtonField
          type="radio"
          name="selected"
          role="option"
          labelText="Yesterday, 1:36PM"
        />
      </Flex>
      <div>
        <Tag tagType="positive">published</Tag>
      </div>
    </Flex>
  </div>
</>
```

- **Featured** - used for featured entities that should be highlighted.

```jsx
<Table>
  <TableHead>
    <TableRow>
      <TableCell>Space</TableCell>
      <TableCell>Updated</TableCell>
      <TableCell>Status</TableCell>
    </TableRow>
  </TableHead>
  <TableBody>
    <TableRow>
      <TableCell>My space</TableCell>
      <TableCell>Oct 29, 2020</TableCell>
      <TableCell>
        <Tag tagType="primary">trial</Tag>
      </TableCell>
    </TableRow>
  </TableBody>
</Table>
```

- **Muted** - this `tagType` is deprecated, please use `tagType` `secondary` instead

### Tag sizes

- **default**

```jsx
<Tag tagType="positive">published</Tag>
```

- **small**

```jsx
<Tag tagType="primary-filled" size="small">
  new
</Tag>
```

## Content recommendations

- try to use labels with short, scannable text
- try to use a single word to describe the status of an element
- try to describe the status in the past tense, like changed or archived

## Accessibility

- text must be clear enough for color-blind users to be able to understand immediately without needing to rely only on the color.
