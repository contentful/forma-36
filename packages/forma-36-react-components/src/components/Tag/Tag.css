@import 'settings/colors';
@import 'settings/typography';
@import 'settings/transitions';
@import 'settings/dimensions';
@import 'settings/borders';

.Tag {
  display: inline-block;
  font-family: var(--font-stack-primary);
  font-weight: var(--font-weight-demi-bold);
  font-size: calc(1rem * (12 / var(--font-base-default)));
  line-height: 20px;
  max-height: 20px;
  text-transform: uppercase;
  letter-spacing: 0.06rem; /*move to tokens or update wide letter spacing token*/
  padding: 0 var(--spacing-xs);
  border-radius: var(--border-radius-small);
}

.Tag--primary {
  color: var(--blue-600);
  background-color: var(--blue-200);
}

.Tag--positive {
  color: var(--green-600);
  background-color: var(--green-200);
}

.Tag--negative {
  color: var(--red-600);
  background-color: var(--red-200);
}

.Tag--warning {
  color: var(--orange-600);
  background-color: var(--orange-200);
}

.Tag--secondary {
  color: var(--gray-700);
  background-color: var(--gray-200);
}

.Tag--muted {
  color: var(--color-text-mid);
  background-color: var(--color-element-lightest);
}

.Tag--primary-filled {
  color: var(--color-white);
  background-color: var(--color-primary);
}

.Tag--featured {
  color: var(--purple-600);
  background-color: var(--purple-200);
}

.Tag--small {
  padding: 3px var(--spacing-2xs);
  font-size: 0.625rem;
  line-height: 0.625rem;
  max-height: 16px;
}
