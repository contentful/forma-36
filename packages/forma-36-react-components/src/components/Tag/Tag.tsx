import React, { Component } from 'react';
import cn from 'classnames';

const styles = require('./Tag.css');

export interface TagProps
  extends React.DetailedHTMLProps<
    React.HTMLAttributes<HTMLDivElement>,
    HTMLDivElement
  > {
  extraClassNames?: string;
  children: React.ReactNode;
  tagType?:
    | 'primary'
    | 'positive'
    | 'negative'
    | 'warning'
    | 'secondary'
    | 'muted';
  testId?: string;
}

export class Tag extends Component<TagProps> {
  static defaultProps = {
    tagType: 'primary',
    testId: 'cf-ui-tag',
  };

  render() {
    const {
      extraClassNames,
      children,
      tagType,
      testId,
      ...otherProps
    } = this.props;

    const classNames = cn(styles.Tag, extraClassNames, {
      [styles[`Tag--${tagType}`]]: tagType,
    });

    return (
      <div className={classNames} data-test-id={testId} {...otherProps}>
        {children}
      </div>
    );
  }
}

export default Tag;
