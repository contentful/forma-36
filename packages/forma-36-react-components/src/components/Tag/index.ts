export * from './Tag';
export { default } from './Tag';
