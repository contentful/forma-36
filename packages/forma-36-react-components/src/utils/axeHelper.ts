import { configureAxe } from 'jest-axe';

const axe = configureAxe({
  rules: {
    region: { enabled: false },
  },
});

export default axe;
