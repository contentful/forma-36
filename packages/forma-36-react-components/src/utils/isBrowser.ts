const isBrowser = () => typeof window !== 'undefined';

export default isBrowser;
