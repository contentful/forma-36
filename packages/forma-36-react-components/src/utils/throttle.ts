const throttle = (delay = 200, fn: Function) => {
  let lastCall = 0;
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  const throttleExec = (...args: any[]) => {
    const now = new Date().getTime();
    if (now - lastCall < delay) {
      return;
    }
    lastCall = now;
    return fn(...args); // eslint-disable-line
  };
  return throttleExec;
};

export default throttle;
