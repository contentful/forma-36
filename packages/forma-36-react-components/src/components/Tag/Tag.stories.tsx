import React from 'react';
import { storiesOf } from '@storybook/react';
import { text, select } from '@storybook/addon-knobs';

import Tag from './Tag';

storiesOf('Components/Tag', module)
  .addParameters({
    propTypes: Tag['__docgenInfo'],
    component: Tag,
  })
  .add('default', () => (
    <Tag
      tagType={select(
        'tagType',
        {
          'Primary (default)': 'primary',
          Positive: 'positive',
          Negative: 'negative',
          Warning: 'warning',
          Secondary: 'secondary',
          Muted: 'muted',
        },
        'primary',
      )}
      entityStatusType={select(
        'Entity Status Type',
        {
          Published: 'published',
          Archived: 'archived',
          draft: 'draft',
          changed: 'changed',
          deleted: 'deleted',
          'None (default)': undefined,
        },
        undefined,
      )}
      className={text('className', '')}
    >
      {text('Children', 'Published')}
    </Tag>
  ));
