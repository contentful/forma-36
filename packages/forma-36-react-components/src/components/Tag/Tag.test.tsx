import React from 'react';
import { render } from '@testing-library/react';

import { axe } from '../../utils/axeHelper';
import { Tag } from './Tag';

it('renders the component', () => {
  const { container } = render(<Tag>Tag</Tag>);

  expect(container.firstChild).toMatchSnapshot();
});

it('renders the component with an additional class name', () => {
  const { container } = render(<Tag className="my-extra-class">Tag</Tag>);

  expect(container.firstChild).toMatchSnapshot();
});

it('renders a "primary" Tag', () => {
  const { container } = render(<Tag tagType="primary">Tag</Tag>);

  expect(container.firstChild).toMatchSnapshot();
});

it('renders a "positive" Tag', () => {
  const { container } = render(<Tag tagType="positive">Tag</Tag>);

  expect(container.firstChild).toMatchSnapshot();
});

it('renders a "negative" Tag', () => {
  const { container } = render(<Tag tagType="negative">Tag</Tag>);

  expect(container.firstChild).toMatchSnapshot();
});

it('renders a "warning" Tag', () => {
  const { container } = render(<Tag tagType="warning">Tag</Tag>);

  expect(container.firstChild).toMatchSnapshot();
});

it('renders a "secondary" Tag', () => {
  const { container } = render(<Tag tagType="secondary">Tag</Tag>);

  expect(container.firstChild).toMatchSnapshot();
});

it('renders a "muted" Tag', () => {
  const { container } = render(<Tag tagType="muted">Tag</Tag>);

  expect(container.firstChild).toMatchSnapshot();
});

it('renders a "published" Tag', () => {
  const { container } = render(
    <Tag entityStatusType="published">Published</Tag>,
  );

  expect(container.firstChild).toMatchSnapshot();
});

it('renders a "draft" Tag', () => {
  const { container } = render(<Tag entityStatusType="draft">Draft</Tag>);

  expect(container.firstChild).toMatchSnapshot();
});

it('renders a "archived" Tag', () => {
  const { container } = render(<Tag entityStatusType="archived">Archived</Tag>);

  expect(container.firstChild).toMatchSnapshot();
});

it('renders a "changed" Tag', () => {
  const { container } = render(<Tag entityStatusType="changed">Changed</Tag>);

  expect(container.firstChild).toMatchSnapshot();
});

it('renders a "deleted" Tag', () => {
  const { container } = render(<Tag entityStatusType="deleted">Deleted</Tag>);

  expect(container.firstChild).toMatchSnapshot();
});

it('has no a11y issues', async () => {
  const { container } = render(<Tag>Tag</Tag>);
  const results = await axe(container);

  expect(results).toHaveNoViolations();
});
