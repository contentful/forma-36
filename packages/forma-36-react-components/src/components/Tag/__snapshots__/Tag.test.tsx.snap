// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders a "archived" Tag 1`] = `
<div
  className="Tag Tag--negative"
  data-test-id="cf-ui-tag"
>
  Archived
</div>
`;

exports[`renders a "changed" Tag 1`] = `
<div
  className="Tag Tag--primary"
  data-test-id="cf-ui-tag"
>
  Changed
</div>
`;

exports[`renders a "deleted" Tag 1`] = `
<div
  className="Tag Tag--negative"
  data-test-id="cf-ui-tag"
>
  Deleted
</div>
`;

exports[`renders a "draft" Tag 1`] = `
<div
  className="Tag Tag--warning"
  data-test-id="cf-ui-tag"
>
  Draft
</div>
`;

exports[`renders a "muted" Tag 1`] = `
<div
  className="Tag Tag--muted"
  data-test-id="cf-ui-tag"
>
  Tag
</div>
`;

exports[`renders a "negative" Tag 1`] = `
<div
  className="Tag Tag--negative"
  data-test-id="cf-ui-tag"
>
  Tag
</div>
`;

exports[`renders a "positive" Tag 1`] = `
<div
  className="Tag Tag--positive"
  data-test-id="cf-ui-tag"
>
  Tag
</div>
`;

exports[`renders a "primary" Tag 1`] = `
<div
  className="Tag Tag--primary"
  data-test-id="cf-ui-tag"
>
  Tag
</div>
`;

exports[`renders a "published" Tag 1`] = `
<div
  className="Tag Tag--positive"
  data-test-id="cf-ui-tag"
>
  Published
</div>
`;

exports[`renders a "secondary" Tag 1`] = `
<div
  className="Tag Tag--secondary"
  data-test-id="cf-ui-tag"
>
  Tag
</div>
`;

exports[`renders a "warning" Tag 1`] = `
<div
  className="Tag Tag--warning"
  data-test-id="cf-ui-tag"
>
  Tag
</div>
`;

exports[`renders the component 1`] = `
<div
  className="Tag Tag--primary"
  data-test-id="cf-ui-tag"
>
  Tag
</div>
`;

exports[`renders the component with an additional class name 1`] = `
<div
  className="Tag my-extra-class Tag--primary"
  data-test-id="cf-ui-tag"
>
  Tag
</div>
`;
